content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -1051,25 +1051,58 @@ A portion of the schema is correlatable if all `array` parent schemas are of typ
-
-Guidelines for working with estimated limits:
-- Adding MaxItems, MaxProperties and MaxLength limits on all data accessed by CEL rules is the best practice.
-- O(n) - For simple rules, it is possible to iterate across a single map/list/string without exceeding the limit, but adding limits on all data accessed by CEL rules is the best practice
-- O(n^2)+ the product of the max lengths usually needs to be <1,000,000.  E.g. 1000 for 2 levels of nesting, 100 for 3 levels of nesting
-- O(n^3) - should generally be avoided
-
-Remember the big-O of a rule takes into account how many times the rule will be executed in addition to
-the data that the rule is working with. If a list inside of a list has a CEL expression that uses `self.all`,
-that rule is O(mn), not O(n). Rules that operate on nested aggregate types (lists inside maps, maps inside 
-maps, and so on) will  likely need to have appropriate limits set on all lists/maps involved before cost 
-validation will pass. Rules that perform string operations or use regular expressions inside of a list or map 
-will also likely need to have limits set in as many places as possible for similar reasons.
+Resource consumption of CEL expressions is limited at compile time by static analysis. If an expression is
+estimated to be prohibitively expensive to execute, it will not be allowed; attempting to apply a CRD 
+containing an expression that is estimated to be too expensive will result in a validation error. A similar 
+system is used at runtime that observes the actions the CEL interpreter takes. If the interpreter executes
+too many instructions, execution of the expression will be halted, and an error will result.
+
+In general, both systems will allow expressions that run in constant/O(1) time; these expressions will
+always take the same amount of time regardless of how large their input is. `self.foo == 1` will be allowed.
+But if `foo` is a string and we instead have `self.foo.contains("someString")`, our expression will take 
+longer to execute depending on how long `foo` is. Another example would be if `foo` was an array, and we
+had an expression `self.foo.all(x, x > 5)`. The cost system will always assume the worst-case scenario if
+a limit on the length of `foo` is not given, and this will happen for anything that can be iterated
+over (lists, maps, etc.). Because of this, it is best to put a limit via `maxItems`, `maxProperties`, and 
+`maxLength` for anything that will be processed in a CEL rule.
+
+For example, given this schema with one rule:
+
+```yaml
+openAPIV3Schema:
+        type: object
+        properties:
+          foo:
+            type: array
+            items:
+              type: string
+            x-kubernetes-validations:
+              - rule: "self.all(x, x.contains('a string')"
+```
+
+The cost system will not allow this rule. Both `foo` and its string items are unbounded in size. If we add
```

Comment:
Instead of "Both `foo` and its string items are unbounded in size." , maybe explain that without max limits, that the computed estimated cost will exceed the max allowed estimated cost?